doc/manage_indices.1
.\" Man page for manage_indices.1
.\" Use the following command to view man page:
.\"
.\"  tbl manage_indices.1 | nroff -man | less
.\"
.TH MANAGE_INDICES 1 "January 20, 2007"
.SH NAME
manage_indices \- A tool to manage indices on Emdros databases
.SH SYNOPSIS
\fBmanage_indices\fR --create [options] database-name
.br
\fBmanage_indices\fR --drop [options] database-name
.br
\fBmanage_indices\fR --help
.br
\fBmanage_indices\fR (-V|--version)
.br
.SH DESCRIPTION

\fBmanage_indices\fR is a tool for creating and dropping indices on an
Emdros database's object types.  If the option "--type
object-type-name" is given, only that object type is affected.  If
this option is not given, all object types in the database are
affected.
.SH OPTIONS
\fBmanage_indices\fR supports the following command-line switches:
.TP 6
.BI \-\-help
show help
.TP
.BI \-V\ ,\ \-\-version
show version
.TP
.BI \-\-create
indices are created
.TP
.BI \-\-drop
indices are dropped
.TP
.TP
.BI \-b\ ,\ \-\-backend \ backend
set database backend to `backend'. Valid values are: For PostgreSQL:
"p", "pg", "postgres", and "postgresql". For MySQL: "m", "my", and
"mysql". For SQLite 2.X.X: "2", "s", "l", "lt", "sqlite", and
"sqlite2". For SQLite 3.X.X: "3", "s3", "lt3", and "sqlite3".
.BI \-t\ ,\ \-\-type " object-type-name"
only the object type \fIobject-type-name\fR is affected
.TP
.BI \-h\ ,\ \-\-host \ hostname 
set db back-end hostname to connect to (default is 'localhost').  Has
no effect on SQLite.
.TP
.BI \-u\ ,\ \-\-user \ user
set database user to connect as (default is 'emdf').  Has no effect on
SQLite.
.TP
.BI \-p\ ,\ \-\-password \ password
set password to use for the database user.  Has no effect on SQLite,
unless you have an encryption-enabled SQLite, in which case this gets
passed as the key.


.SH RETURN VALUES
.TP
.BR 0 " Success"
.TP
.BR 1 " Wrong usage"
.TP
.BR 2 " Connection to backend server could not be established"
.TP
.BR 3 " An exception occurred (the type is printed on stderr)"
.TP
.BR 5 " Database error"
.SH AUTHORS
Written Ulrik Sandborg-Petersen (ulrikp@emdros.org).


NEWS
This file is for release-notes for Emdros.

- *** Version 3.7.0 *** 

Release notes for Emdros 3.7.0
==============================

- Emdros has been relicensed under the MIT license.

- All copyright in Emdros and related documentation and build scripts
  has been assigned to Ulrik Sandborg-Petersen's holding company,
  Sandborg-Petersen Holding ApS.  This is so as to be able to serve
  customers better with support and other Emdros-related help.


- *** Version 3.6.1 *** 

Release notes for Emdros 3.6.1
==============================

Emdros version 3.6.1 is a maintenance release with the following
summary of changes since 3.6.0:

- The MQL query language has been extended with the ability to compare
  features with type LIST OF INTEGER, LIST OF ID_D, and LIST OF
  <enum-type> with an explicit list (including the empty list) using
  the "=" comparison operator.

  For example:

  SELECT ALL OBJECTS
  WHERE
  [Clause_atom
     daughters = () // Test for empty list
     OR
     daughters = (12345,67891) // Test for explicit, non-empty list
  ]
  GO

- Detection of PHP 5 and PHP 7 on macOS now works correctly.

- Compileability on Solaris 10 with Sun's CC C++ compiler has been
  restored.

- Compatibility with wxWidget 2.8.12 has been restored.


Enjoy!


Ulrik Sandborg-Petersen



- *** Version 3.6.0 *** 

Release notes for Emdros 3.6.0
==============================

The release notes below cover some of the important changes in Emdros
since 3.5.0.

- SQLite2 support has been removed, as version 2 of SQLite has not
  been updated by the SQLite authors for many, many years.  Also,
  SQLite2 support in Emdros has been been deprecated for some time.

- The MQL subsystem has a better query planner, which is able to
  choose the "Innermost First" strategy, rather than the default
  "Outermost First" strategy, under many more circumstances.

- The Emdros MySQL backend now has support for transactions, if
  available in the MySQL server.

- Assorted performance improvements to the following subsystems:

  - XML parsing (renderobjects, for example)
  - MQL

- Node.JS support has been removed.  The reason is that the latest
  SWIG version (3.0.12) does not support the latest Node.  Also, the
  Node.JS bindings never really worked.

- PCRE has been upgraded.

- SQLite3 has been upgraded.

- JSON output now escapes string characters correctly.

- Compiling Linux .rpm packages works on recent systems (CentOS 7,
  Fedora 28).

- Compiling Linux .deb packages works on recent systems (Ubuntu 16.04,
  Ubuntu 18.04, Debian 9.4).

- Building on Windows supports the latest Visual Studio.

- Lots of compiler warnings have been removed.

- the harvest template language has new constructs:

  - {{ inccounter <string> var <string> }}
  - {{ deccounter <string> var <string> }}

- The Full Text Search FTS3 subsystem is now able to obey
  largest_proximity, and to use * and ? in the query terms.

- A new program was added, bptdetach(1), which enables detaching
  "payloads" from BPT databases.

- The BPT backend (if available -- ask Ulrik for the source code
  licensing terms if you are interested) now supports LZMA compression
  of payloads.

- Source code using wxWidgets has been updated to support wxWidgets
  2.8, 3.0, and 3.1.

- The Emdros PostgreSQL backend now adds "TEMPLATE template0" when
  doing CREATE DATABASE.  This is so as to avoid using template1,
  which may conflict with the encoding used (and does so on Debian
  9.4).

- The amalgamation output has been split into several files.  This is
  because the Android NDK MIPS target cannot compile the old,
  monolithic amalgamation files.



- *** Version 3.5.0 *** 

Release notes for Emdros 3.5.0
==============================

The focus-areas of this release include:

- Enhancements to the core MQL
- Bugfixes to the core MQL.
- Enhancements to the render library
- Enhancements to the Full Text Search subsystem
- Enhancements to SWIG bindings
- Enhancements to the build system

Please consult the ChangeLog (going back to version 3.4.1.pre01) for
the specifics.  You can also consult the MQL Programmer's Guide for
the enhancements to the core MQL, and the Programmer's Reference Guide
for the enhancements to the Render library and the FTS subsystem.

Briefly, the core MQL has been enhanced as follows:

- SELECT ALL OBJECTS has a new WITH MAX RANGE clause, helping to limit
  otherwise runaway queries such as [Word]..[Word].

- SELECT ALL OBJECTS now sports an [UnorderedGroup] block.

- GET AGGREGATE FEAUTURES is a new statement, similar to SQL's MIN(),
  MAX(), SUM(), and COUNT() aggregate functions.


- *** Version 3.4.0 *** 

Release notes for Emdros 3.4.0
==============================

The focus-areas of this release include:

- Support for iOS and Android
- A smoother user experience for the GUI applications
- Enhancements to the MQL query language
- Documentation improvements
- Build improvements on all platforms
- Speed improvements


Binaries for Windows(R) and Mac OS X(R)
---------------------------------------

As usual, binaries are available for Mac OS X and Windows(R).

The Windows binaries are available in both x86 and x64 versions.  They
have built-in support for MySQL, SQLite 2, and SQLite 3.  The x86
binaries are built with Visual Studio Express 2010.  The x64 binaries
are built with Windows SDK 6.0A.

The Mac OS X binaries are Universal binaries (i386 and x86_64) running
on Mac OS X 10.6 (Snow Leopard), 10.7 (Lion), and 10.8 (Mountain
Lion), 10.9 (Mavericks), and possibly later versions of Mac OS X.
They do not have support for either MySQL or PostgreSQL; Only SQLite 2
and SQLite 3 are supported in the Mac OS X binaries.  You might be
able to compile the sources with support for MySQL yourself, and
possibly also PostgreSQL.

Linux binaries

src/manage_indices.cpp
/*
 * manage_indices.cpp
 *
 *
 *
 * Tool to manage indices on Emdros databases.  Either creates or
 * drops all indices.  If a "--type object-type" argument is given,
 * then only that object type gets its indices created or dropped.  If
 * no such argument is given, all all object types in the database
 * have their indices created or dropped.
 *
 * Synopsis:
 *   manage_indices --create [options] database-name
 *   manage_indices --drop [options] database-name
 *   manage_indices --help
 *
 * Supports -t, -h, -u, and -p arguments (see output from --help).
 * 
 *
 *
 * Created: 11/13-2001
 * Last update: 5/11-2018
 *
 * Return codes:
 * 
 * 0 : Success
 * 1 : Wrong usage
 * 2 : Connection to backend server could not be established
 * 3 : An exception occurred
 * 5 : Database error
 * */

/*
 * Copyright (C) 2001-2018     Ulrik Sandborg-Petersen
 * Copyright (C) 2018-present  Sandborg-Petersen Holding ApS, Denmark
 *
 * Licensed under the MIT License.
 *
 * Please see the file COPYING in the root of the sources for more details.
 *
 */



#include <emdfdb.h>
#include <iostream>
#include <sstream>
#include <fstream>
#include <string>
#include <infos.h>
#include <monads.h>
#include <utils.h>
#include <emdf_output.h>
#include <encryption.h>
#include <debug.h>
#include <opt.h>
#include <messages.h>



#if USE_POSTGRESQL
#include <pgemdfdb.h>
#endif
#if USE_MYSQL
#include <mysqlemdfdb.h>
#endif
#if USE_SQLITE2
#include <sqliteemdfdb.h>
#endif
#if USE_SQLITE3
#include <sqlite3emdfdb.h>
#endif


#define TASK_NONE    (0)
#define TASK_CREATE  (1)
#define TASK_DROP    (2)

void show_version(std::ostream& ostr)
{
	ostr << "manage_indices from Emdros version " << EMDROS_VERSION << std::endl;
}

void print_usage(std::ostream& ostr)
{
	show_version(ostr);
	ostr << "Usage:" << std::endl;
	ostr << "    manage_indices  --create  [options]  database-name  " << std::endl;
	ostr << "    manage_indices  --drop    [options]  database-name" << std::endl;
	ostr << "    manage_indices  --help" << std::endl;
	ostr << "    manage_indices  (-V|--version)" << std::endl;
	ostr << "OPTIONS:" << std::endl;
	printUsageStandardArguments(ostr);
	ostr << "   -t, --type objecttype" << std::endl;

	ostr << "DEFAULTS:" << std::endl;
	printUsageDefaultsOfStandardArguments(ostr);
}


int main(int argc, char* argv[])
{
	EMdFDB* pDB;

	int nTask = TASK_NONE;
	bool bObjectType = false;
	std::string strObjectType;
	std::string strDatabase;
	std::string hostname;
	std::string user;
	std::string password;
	eCharsets dummy_charset;
	eBackendKind backend_kind;

	addStandardArguments(false // Do not add -e | --encoding
			     );
	addOption("--create", "--create");
	addOption("--drop", "--drop");
	addOption("-t", "--type", true,
		  "",
		  "ERROR:  -t and --type must have an existing object type as their argument.\n"
		  "       example: -t clause\n");

	// Parse arguments
	std::list<std::string> surplus_arguments;

	std::string error_message;
	if (!parseArguments(argc, argv, error_message, surplus_arguments)) {
		print_usage(std::cerr);
		std::cerr << error_message << std::endl;
		return 1;
	} else {
		bool bShowVersion;
		bool bShowHelp;
		std::string error_message;
		if (!getStandardArguments(bShowVersion, bShowHelp,
					  hostname,
					  user,
					  password,
					  backend_kind,
					  dummy_charset,
					  error_message)) {
			print_usage(std::cerr);
			std::cerr << error_message << std::endl;
			return 1;
		}

		// -V | --version
		if (bShowVersion) {
			show_version(std::cout);
			return 0;
		}

		// --help
		if (bShowHelp) {
			print_usage(std::cout);
			return 0;
		}

		// task
		if (getArgumentPresent("--create")) {
			nTask = TASK_CREATE;
		}
		if (getArgumentPresent("--drop")) {
			nTask = TASK_DROP;
		} 

		if (nTask != TASK_DROP
		    && nTask != TASK_CREATE) {
			print_usage(std::cerr);
			std::cerr << "ERROR: You must specify one of --create or --drop when not doing\n"
				  << "       --version or --help." << std::endl;
			return 1;
		}

		bObjectType = getArgumentValue("-t", strObjectType);
		if (bObjectType
		    && strObjectType == "") {
			print_usage(std::cerr);
			std::cerr << "ERROR: -t and --type must have an object type as their argument,\n"
				  << "       not the empty string!" << std::endl;
			return 1;
		}

		if (surplus_arguments.size() == 0) {
			strDatabase = "";
			print_usage(std::cerr);
			std::cerr << "ERROR: You must specify a database on the command line." << std::endl;
			return 1;
		} else if (surplus_arguments.size() == 1) {
			strDatabase = *surplus_arguments.begin();
		} else {
			print_usage(std::cerr);
			std::cerr << "Error: You can only specify one database. You specified: '" << joinList("' and '", surplus_arguments) << "'." << std::endl;
			return 1;
		}
	}
	
	
	// Make connection
	if (backend_kind == kPostgreSQL) {
#if USE_POSTGRESQL
		pDB = new PgEMdFDB(hostname, user, password, strDatabase);
#else
		throw EmdrosException(std::string("PostgreSQL support not compiled in. Cannot make a PostgreSQL connection."));
#endif
	} else if (backend_kind == kMySQL) {
#if USE_MYSQL
		pDB = new MySQLEMdFDB(hostname, user, password, strDatabase);
#else
		throw EmdrosException(std::string("MySQL support not compiled in. Cannot make a MySQL connection."));
#endif
	} else if (backend_kind == kSQLite2) {
		throw EmdrosException(std::string("SQLite 2.X.X is no longer supported."));
	} else if (backend_kind == kSQLite3) {
#if USE_SQLITE3
		pDB = new SQLite3EMdFDB(strDatabase, password);
#else
		throw EmdrosException(std::string("SQLite 3.X.X support not compiled in. Cannot make a SQLite 3.X.X connection."));
#endif
	} else {
		print_usage(std::cerr);
		std::cerr << "Unknown backend." << std::endl;
		return 1;
	}


	// Check that we connected
	if (!pDB->connectionOk()) {
		emdrosMessageConnectionNotOK(&std::cerr, backend_kind);
		delete pDB;
		return 2;
	}

	int nResult = 3;
	try {
		switch (nTask) {
		case TASK_CREATE:
			if (bObjectType) {
				nResult = (pDB->createIndicesOnObjectType(strDatabase, password, strObjectType)) ? 0 : 5;
			} else {
				nResult = (pDB->createIndicesOnDatabase(strDatabase, password)) ? 0 : 5;
			}
			break;
		case TASK_DROP:
			if (bObjectType) {
				nResult = (pDB->dropIndicesOnObjectType(strDatabase, password, strObjectType)) ? 0 : 5;
			} else {
				nResult = (pDB->dropIndicesOnDatabase(strDatabase, password)) ? 0 : 5;
			}
			break;
		default:
			ASSERT_THROW(false, "Unknown task kind");
			break;
		}
		if (nResult != 0) {
			nResult = 5;
			std::cerr << "ERROR: Database error..." << std::endl;
			std::cerr << pDB->errorMessage() << std::endl; 
		}
	} catch (EMdFDBException& e) {
		std::cerr << "ERROR: EMdFDBException (Database error)..." << std::endl;
		std::cerr << pDB->errorMessage() << std::endl; 
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (BadMonadsException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (WrongCharacterSetException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFOutputException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EmdrosException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (...) {
		std::cerr << "Unknown exception occurred.  Program aborted." << std::endl;
	} 

	// Zero-fill password
	zeroFillString(password);

	// Clean up
	delete pDB;
	return nResult;
}


